Create rule S6680(C#): Loop boundaries should not be vulnerable to injection attacks (#2607)
SonarSource/rspec: rules/S6680/csharp/rule.adoc

This vulnerability exposes the system to various operational overloads that can
lead to either a technical denial of service and/or business disruptions.

== Why is this an issue?

In the context of an application vulnerable to Loop Boundary Injection: +
After discovering the injection point, attackers insert data into the
vulnerable field to either make the affected component inaccessible, attempt a
malfunction, or read from an artifact that exceeds the developer's intended
boundaries.

In low-level languages, this vulnerability can lead to memory over-reads or

> over-writes and have worse consequences.

Reviewer: So here there was a mistake from my side, but also a small comment. 

1- This text is mistakenly in the description instead of the impact

2- The sentence structure "In the context of an application vulnerable to Loop Boundary Injection: +
After discovering the injection point, attackers" is actually there for consistency across vulns, we write the same everywhere. If you think we should change that because that feels weird, we should do something bigger at the squad level

> over-writes and have worse consequences.

Reviewer: So I am going to keep some part and take some of yours 👍

=== What is the potential impact?

==== Self Denial of service

If the component jeopardized by this vulnerability is not a bottleneck that
acts as a single point of failure (SPOF) within the application, the denial of
service might only affect the attacker who initiated it.

> Such benign denial of service can also occur in architectures that rely heavily
> on containers and container orchestrators. Replication systems would detect the
> failure of a container and automatically replace it.

Reviewer 2: Again, I can see what you're saying here but I think that the language could be improved to make it clearer for a junior developer.

```suggestion
Even if the denial of service has little direct impact, it can cause secondary
effects in architectures that use containers and container orchestrators. The
increased CPU usage could be seen as a sign that additional containers need to
be started to handle increasing load. It could also be seen as a sign that the
container is failing, causing it to be replaced prematurely.
```

Reviewer: Here I am not a fan of talking about CPU usage as signs, but it's true that the simpler language makes that part less elitist. I'll take your text. I think I will reduce the parts about replication, it seems overkill

----
public class ExampleController : Controller
{
    public IActionResult Compute(int data)
    {
        for (int i = 0; i < data; i++) // Noncompliant
        {
            Console.WriteLine("Hello");
        }

        Enumerable
            .Range(1, data) // Noncompliant
            .ToList()
            .ForEach(i => Console.WriteLine("World"));

        return Ok();
    }
}
----

==== Compliant solution

[source,csharp,diff-id=1,diff-type=compliant]
----
public class ExampleController : Controller
{
    public static int MAX_BOUNDARY = 1337;
    public static int MIN_BOUNDARY = 1;

    public IActionResult Compute(int data)
    {
        
        if (MIN_BOUNDARY > data)
        {
            data = MIN_BOUNDARY;
        }
        else if (data > MAX_BOUNDARY)
        {
            data = MAX_BOUNDARY;
        }

        for (int i = 0; i < data; i++)
        {
            Console.WriteLine("Hello");
        }

        Enumerable
            .Range(1, data) // Noncompliant
            .ToList()
            .ForEach(i => Console.WriteLine("World"));

        return Ok();
    }
}
----

=== How does this work?

==== Set limits

> just like with other injection vulnerabilities, when fixing this vulnerability
> you need to make sure that the value to be set as a boundary respects the
> system and functional boundaries. If an instruction should not be executed more
> than 3 times, the limit must be explicitly set. +
> No assumptions should be made about what users will choose, because attackers
> definitely do not adhere to these assumptions.

Reviewer 2: What you're trying to say here is good, but the choice of wording makes it fairly obvious that you're not a native English speaker.

Turning again to ChatGPT for inspiration:
```suggestion
Validate loop variable values to ensure they fall within the expected bounds.
If a value falls outside of these bounds, reject it as invalid or adjust it to
bring it within bounds. This guarantees that the loop will only iterate over
the expected number of elements or perform the expected number of iterations.

Do not assume that users will provide sensible values. Attackers intentionally
choose unusual values to cause the system to misbehave.
```

== Resources

=== Standards

* https://owasp.org/Top10/A03_2021-Injection/[OWASP Top 10 2021 Category A3] - Injection
* https://www.owasp.org/index.php/Top_10-2017_A1-Injection[OWASP Top 10 2017 Category A1] - Injection
* https://cwe.mitre.org/data/definitions/606[MITRE, CWE-606] - Unchecked Input for Loop Condition

> 

Reviewer: ```suggestion

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

Change this code to not construct class|method names directly from user-controlled data.

'''
endif::env-github,rspecator-view[]
```

This vulnerability exposes the system to various operational overloads that can
lead to either a technical denial of service and/or business disruptions.

== Why is this an issue?

> In the context of an application vulnerable to Loop Boundary Injection: +
> After discovering the injection point, attackers insert data into the
> vulnerable field to either make the affected component inaccessible, attempt a
> malfunction, or read from an artifact that exceeds the developer's intended
> boundaries.
> 
> In low-level languages, this vulnerability can lead to memory over-reads or
> over-writes and have worse consequences.

Reviewer: ```suggestion
Loop boundary injections occur in an application when the application retrieves
data from a user or a third-party service and inserts it into a loop or a function
acting as a loop, without sanitizing it first.

If an application contains a loop that is vulnerable to injections, 
it is exposed to attacks that target its availability where that loop is used.

A user with malicious intent carefully performs actions whose goal is to cause the 
loop to run for more iterations than the developer intended, resulting in unexpected
behavior or even a crash of the program.

After creating the malicious request, the attacker can attack the servers
affected by this vulnerability without relying on any prerequisites.
```

=== What is the potential impact?

> 

Reviewer: ```suggestion
In the context of an application vulnerable to Loop Boundary Injection: +
After discovering the injection point, attackers insert data into the
vulnerable field to either make the affected component inaccessible, attempt a
malfunction, or read from an artifact that exceeds the developer's intended
boundaries.

In languages that don't enforce memory access checks, this can also lead to a buffer
overflow or underflow which may result in sensitive information disclosure or
remote code execution.

```

==== Self Denial of service

If the component affected by this vulnerability is not a bottleneck that
acts as a single point of failure (SPOF) within the application, the denial of
service might only affect the attacker who initiated it.

> Such benign denial of service can also occur in architectures that rely heavily
> on containers and container orchestrators. Replication systems would detect the
> failure of a container and automatically replace it.

Reviewer: ```suggestion
Even if the denial of service has little direct impact, it can cause secondary
effects in architectures that use containers and container orchestrators. It
could cause unexpected container failures or resource overconsumption,
for example.
```

        }

        Enumerable
            .Range(1, data) // Noncompliant
            .ToList()
            .ForEach(i => Console.WriteLine("World"));

        return Ok();
    }
}
----

=== How does this work?

==== Set limits

Validate loop variable values to ensure they fall within the expected bounds.
If a value falls outside of these bounds, reject it as invalid or adjust it to
bring it within bounds. This guarantees that the loop will only iterate over
the expected number of elements or perform the expected number of iterations.

Do not assume that users will provide sensible values. Attackers intentionally
choose unusual values to cause the system to misbehave.

== Resources

=== Standards

* https://owasp.org/Top10/A03_2021-Injection/[OWASP Top 10 2021 Category A3] - Injection
* https://www.owasp.org/index.php/Top_10-2017_A1-Injection[OWASP Top 10 2017 Category A1] - Injection
* https://cwe.mitre.org/data/definitions/606[MITRE, CWE-606] - Unchecked Input for Loop Condition

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

Change this code to set loop bounds directly from user-controlled data.

> 

Reviewer: ```suggestion

=== Highlighting

"[varname]" is tainted (assignments and parameters)

this argument is tainted (method invocations)

the returned value is tainted (returns & method invocations results)

```

behavior or even a crash of the program.

After creating the malicious request, the attacker can attack the servers
affected by this vulnerability without relying on any prerequisites.

=== What is the potential impact?
In the context of an application vulnerable to Loop Boundary Injection: +
After discovering the injection point, attackers insert data into the
vulnerable field to either make the affected component inaccessible, attempt a
malfunction, or read from an artifact that exceeds the developer's intended
boundaries.

In languages that don't enforce memory access checks, this can also lead to a buffer
overflow or underflow which may result in sensitive information disclosure or
remote code execution.

==== Self Denial of service

If the component affected by this vulnerability is not a bottleneck that
acts as a single point of failure (SPOF) within the application, the denial of
service might only affect the attacker who initiated it.

Even if the denial of service has little direct impact, it can cause secondary
effects in architectures that use containers and container orchestrators. It
could cause unexpected container failures or resource overconsumption,
for example.

> include::../../../shared_content/appsec/high_impact_dos.adoc[]

Reviewer: ```suggestion
==== Infrastructure SPOFs

However, a denial of service attack can be critical to the enterprise if it
targets a SPOF component. Sometimes the SPOF is a software architecture
vulnerability (such as a single component on which multiple critical components
depend) or an operational vulnerability (for example, insufficient container
creation capabilities or failures from containers to terminate).

In either case, attackers aim to exploit the infrastructure weakness by sending
as many malicious payloads as possible, using potentially huge offensive
infrastructures.

These threats are particularly insidious if the attacked organization does not
maintain a disaster recovery plan (DRP).
```

This vulnerability exposes the system to various operational overloads that can
lead to either a technical denial of service and/or business disruptions.

== Why is this an issue?

Loop boundary injections occur in an application when the application retrieves
data from a user or a third-party service and inserts it into a loop or a function
acting as a loop, without sanitizing it first.

If an application contains a loop that is vulnerable to injections, 
it is exposed to attacks that target its availability where that loop is used.

A user with malicious intent carefully performs actions whose goal is to cause the 
loop to run for more iterations than the developer intended, resulting in unexpected
behavior or even a crash of the program.

After creating the malicious request, the attacker can attack the servers
affected by this vulnerability without relying on any prerequisites.

=== What is the potential impact?

> In the context of an application vulnerable to Loop Boundary Injection: +

Reviewer: Yeah I agree. This was added after discussions but I agree on removing it

creation capabilities or failures from containers to terminate).

In either case, attackers aim to exploit the infrastructure weakness by sending
as many malicious payloads as possible, using potentially huge offensive
infrastructures.

These threats are particularly insidious if the attacked organization does not
maintain a disaster recovery plan (DRP).

== How to fix it

=== Code examples

==== Noncompliant code example

[source,csharp,diff-id=1,diff-type=noncompliant]
----
public class ExampleController : Controller
{
    public IActionResult Compute(int data)
    {
        for (int i = 0; i < data; i++) // Noncompliant
        {
            Console.WriteLine("Hello");
        }

        Enumerable
            .Range(1, data) // Noncompliant
            .ToList()
            .ForEach(i => Console.WriteLine("World"));

        return Ok();
    }
}
----

==== Compliant solution

[source,csharp,diff-id=1,diff-type=compliant]
----
public class ExampleController : Controller
{
    public static int MAX_BOUNDARY = 1337;
    public static int MIN_BOUNDARY = 1;

    public IActionResult Compute(int data)
    {
        

>         if (MIN_BOUNDARY > data)

Reviewer: @jamie-anderson-sonarsource do you think as a C# dev you would have done that ? what is the most performance worthy between that and Min()/Max() functions ?

Reviewer 2: I've seen (and used) both forms of ensuring that a parameter value is within specific limits:
```csharp
int value;
if (value < MIN_ALLOWED)
    value = MIN_ALLOWED;
if (value > MAX_ALLOWED)
    value = MAX_ALLOWED);
```
```csharp
int value;
value = Math.Max(value, MIN_ALLOWED);
value = Math.Min(value, MAX_ALLOWED);
```
The second form means fewer lines of code, especially if there are rules in place to say that the body of an `if` statement needs to have curly braces around it. That would turn the 4 lines of `if` code in my examples into 6 or 8 lines, depending on whether the opening brace is supposed to go on its own line.

It's also quite common to see code like:
```csharp
for (int i = 0; i < Math.Min(value, MAX_ALLOWED); i++)
{
    // ...
}
```
This works because the initial value of `i` is the lower boundary, and the upper boundary cannot go above `MAX_ALLOWED`.
